### PR TITLE
Fixes #24902 - pass in list of packages when publishing CV

### PR DESCRIPTION
**Pull request by beav**
Previously, the only way to control which packages land in content
view version was by using filters, or by managing the contents of the
Library versions of the repos.

This commit lets you optionally specify the exact package set you want
in your repository. It will look at the Library version of each repo,
and copy the list of packages into the CV version.

Example value for `repos_units` when calling
`/katello/api/v2/content_views/2/publish`:

```json
[
    {
        "repo_label": "zoo",
        "rpm_filenames": [
            "walrus-5.21-1.noarch.rpm",
            "gorilla-0.62-1.noarch.rpm"
        ]
    },
    {
        "repo_label": "a_longer_label",
        "rpm_filenames": [
            "facter-2.4.6-3.el7sat.x86_64.rpm",
            "pulp-rpm-handlers-2.13.4.9-1.el7sat.noarch.rpm"
        ]
    }
]
```

This works the same for custom and RH repos.

When used with #7594, you can
set the CV version number as well as the list of packages in each
repo. This is useful for cloning a CV version from one Katello to
another.

### CODE REVIEW

**Comment by beav** [comment]
@ehelms @chris1984 this is the WIP commit for setting package list on publish, it still needs tests and I need to test what happens with composites.

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 1]
```diff
@@ -79,13 +79,16 @@ def update
     param :description, String, :desc => N_("Description for the new published content view version")
     param :force_yum_metadata_regeneration, :bool, :desc => N_("Force yum metadata regeneration on the repositories " \
                                                            "in the content view version")
-    param :major, :number, :desc => N_("Override the major version number"), :required => false
-    param :minor, :number, :desc => N_("Override the minor version number"), :required => false
+    param :major, :number, :desc => N_("Override the major version number")
+    param :minor, :number, :desc => N_("Override the minor version number")
+    param :repos_units, Hash, :desc => N_("list of repositories and units to publish")
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_views_controller.rb]: You should be able to describe this hash in more detail as part of the param definition (showing the structure)

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 2]
```diff
@@ -79,13 +79,16 @@ def update
...
     def publish
+      repos_units = params[:repos_units] ? JSON.parse(params[:repos_units]) : nil
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_views_controller.rb]: why is the JSON.parse necessary?

File: app/lib/actions/katello/repository/clone_to_version.rb [hunk 3]
```diff
@@ -5,13 +5,16 @@ class CloneToVersion < Actions::Base
         # allows accessing the build object from the superior action
         attr_accessor :new_repository
 
-        def plan(repositories, content_view_version, incremental = false)
+        def plan(repositories, content_view_version, options)
+          incremental = options[:incremental] ? true : false
```
Comment (jlsherrill) [inline app/lib/actions/katello/repository/clone_to_version.rb]: `incremental = options.fetch(:incremental, false)` would be preferrable here

Comment (jlsherrill) [reply app/lib/actions/katello/repository/clone_to_version.rb]: and you probably want to make options optional `options = {}`

File: app/lib/actions/katello/repository/clone_to_version.rb [hunk 4]
```diff
@@ -5,13 +5,16 @@ class CloneToVersion < Actions::Base
...
           content_view = content_view_version.content_view
           filters = incremental ? [] : content_view.filters.applicable(repositories.first)
 
           self.new_repository = repositories.first.build_clone(content_view: content_view,
                                                                version: content_view_version)
 
+          repo_units = options[:repos_units] ? options[:repos_units]["repos"].detect { |r| r["label"] == repositories.first.label } : nil
```
Comment (jlsherrill) [inline app/lib/actions/katello/repository/clone_to_version.rb]: ```           repo_units = options[:repos_units]["repos"].detect { |r| r["label"] == repositories.first.label } if options[:rpeo_units]```

Would be preferable

**Comment by jlsherrill** [comment]
whoops i just saw this was WIP, a few nitpicks, otherwise looks good

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 2 above)
Comment (beav) [reply app/controllers/katello/api/v2/content_views_controller.rb]: @jlsherrill without the parse, it will just keep the field as a string and not let me access the fields stored inside easily. I'm not sure if this is expected behavior or not.

Comment (jlsherrill) [reply app/controllers/katello/api/v2/content_views_controller.rb]: hrmmm, you sure its being passed to the api correctly?   what does the request look like?

Comment (beav) [reply app/controllers/katello/api/v2/content_views_controller.rb]: example: `http -j -a admin:changeme --verify=no POST https://localhost/katello/api/v2/content_views/4/publish major=5 minor=25 repos_units=@packages.json`

I confirmed that it's sending `Content-Type: application/json` in the header, I feel like this should work without calling `JSON.parse` but not sure what's happening.

Comment (jlsherrill) [reply app/controllers/katello/api/v2/content_views_controller.rb]: i think apipie there is sending repo_units as a string, such as:

{ "repo_units": "[\"this should be json, but is actually a string\"]" }

**Comment by beav** [comment]
TODO: a couple more tests, and figure out apipie issue

**Comment by beav** [comment]
@jlsherrill the issue requiring `JSON.parse()` was that I was passing `repos_units=@packages.json` but I needed to use `repos_units:=@packages.json`.

**Comment by beav** [comment]
@ehelms @chris1984 I made an update that breaks `import.sh` and `export.sh`, the new format for `repos_units` is like so:

```json
[
  {
    "repo_label": "zoo",
    "rpm_filenames": [
      "walrus-5.21-1.noarch.rpm",
      "gorilla-0.62-1.noarch.rpm"
    ]
  },
  {
    "repo_label": "Red_Hat_Satellite_Tools_6_3_for_RHEL_7_Server_RPMs_x86_64",
    "rpm_filenames": [
      "facter-2.4.6-3.el7sat.x86_64.rpm",
      "pulp-rpm-handlers-2.13.4.9-1.el7sat.noarch.rpm"
    ]
  }
]
```

**Comment by beav** [comment]
[test prprocessor]

**Comment by beav** [comment]
:greenland:

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 5]
```diff
@@ -79,13 +79,25 @@ def update
...
     param :force_yum_metadata_regeneration, :bool, :desc => N_("Force yum metadata regeneration on the repositories " \
                                                            "in the content view version")
-    param :major, :number, :desc => N_("Override the major version number"), :required => false
-    param :minor, :number, :desc => N_("Override the minor version number"), :required => false
+    param :major, :number, :desc => N_("Override the major version number")
+    param :minor, :number, :desc => N_("Override the minor version number")
+
+    param :repos_units, Hash, :desc => N_("Specify the list of units in each repo") do
+      param :repo_label, String, :required => true
+      param :rpm_filenames, Array, :required => true do
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_views_controller.rb]: How does one specify errata?

File: app/lib/actions/katello/repository/clone_to_version.rb [hunk 6]
```diff
@@ -38,6 +41,13 @@ def plan(repositories, content_view_version, incremental = false)
             end
           end
         end
+
+        def extract_rpm_filenames(repos_units, repo_label)
+          if repos_units.present?
```
Comment (ehelms) [inline app/lib/actions/katello/repository/clone_to_version.rb]: You could just return here:

```
return unless repo_units.present?
```

File: app/lib/actions/katello/repository/clone_to_version.rb [hunk 7]
```diff
@@ -38,6 +41,13 @@ def plan(repositories, content_view_version, incremental = false)
...
+            repo_units = repos_units.detect { |r| r[:repo_label] == repo_label }
```
Comment (ehelms) [inline app/lib/actions/katello/repository/clone_to_version.rb]: This is to double check that the passed in repo_label is present within the units?

File: app/lib/actions/katello/repository/clone_yum_content.rb [hunk 8]
```diff
@@ -19,6 +21,16 @@ def plan(source_repo, target_repo, filters, purge_empty_units, options = {})
...
           end
 
+          # if we are providing a list of files, process that here and override filters
+          if rpm_filenames.present?
+            # ensure we have all the files we want to copy
+            rpms_available = source_repo.rpms.pluck(:filename)
+            rpm_filenames.each do |filename|
+              fail "%s not available in repository %s" % [filename, source_repo.label] unless rpms_available.include? filename
+            end
+            copy_clauses = { 'filename' => { '$in' => rpm_filenames } }
```
Comment (ehelms) [inline app/lib/actions/katello/repository/clone_yum_content.rb]: The intent here is to override the copy_clauses from above if any are defined? That might be unexpected behavior to a user if they don't get a warning like "you cannot provide filters and a list of RPMs"

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 9]
```diff
@@ -79,13 +79,25 @@ def update
     param :description, String, :desc => N_("Description for the new published content view version")
     param :force_yum_metadata_regeneration, :bool, :desc => N_("Force yum metadata regeneration on the repositories " \
                                                            "in the content view version")
-    param :major, :number, :desc => N_("Override the major version number"), :required => false
-    param :minor, :number, :desc => N_("Override the minor version number"), :required => false
+    param :major, :number, :desc => N_("Override the major version number")
+    param :minor, :number, :desc => N_("Override the minor version number")
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_views_controller.rb]: We sure we want to change this to require major and minor? That may break the UI and/or dependent API projections (e.g. Ansible publish module)

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 10]
```diff
@@ -79,13 +79,25 @@ def update
...
+
+    param :repos_units, Hash, :desc => N_("Specify the list of units in each repo") do
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_views_controller.rb]: Should this be required false?

File: app/lib/actions/katello/content_view/publish.rb [hunk 11]
```diff
@@ -9,11 +9,26 @@ class Publish < Actions::EntryAction
...
+            labels_from_repos_units = options[:repos_units].map { |l| l[:repo_label] }
+
+            labels_from_repos_units.each do |l|
+              fail _("Repository label '%s' is not associated with content view.") % l unless valid_labels_from_cv.include? l
+            end
+
+            valid_labels_from_cv.each do |l|
+              fail _("Content view has repository label '%s' which is not specified in repos_units parameter.") % l unless labels_from_repos_units.include? l
+            end
+          end
```
Comment (ehelms) [inline app/lib/actions/katello/content_view/publish.rb]: Can we please avoid one letter variable names. Even within a simple map or each statement, the readability is decreased with their use.

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 9 above)
Comment (chris1984) [reply app/controllers/katello/api/v2/content_views_controller.rb]: Had to change this to :required => false so hammer would be happy and let me publish, after talking to @ehelms api/webui will not care about this but hammer hard fails with required off.

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 10 above)
Comment (chris1984) [reply app/controllers/katello/api/v2/content_views_controller.rb]: Had to change this to :required => false so hammer would be happy and let me publish, after talking to @ehelms api/webui will not care about this but hammer hard fails with required off.

Comment (beav) [reply app/controllers/katello/api/v2/content_views_controller.rb]: @chris1984 ah interesting, thx

File: app/lib/actions/katello/repository/clone_yum_content.rb (hunk 8 above)
Comment (beav) [reply app/lib/actions/katello/repository/clone_yum_content.rb]: At least one user wanted a true "override" that ignored filters, so I added a log message when someone has filters + pkglist overrides set.

**Comment by beav** [comment]
@ehelms @chris1984 this pr should be good to go again

**Comment by chris1984** [comment]
@beav sounds good I will test again and report back.

**Comment by beav** [comment]
@ehelms @jlsherrill ok to merge? I still need ack from someone with write access to unlock the magic merge button.

File: app/lib/actions/katello/content_view/publish.rb [hunk 12]
```diff
@@ -9,11 +9,26 @@ class Publish < Actions::EntryAction
         def plan(content_view, description = "", options = {})
           action_subject(content_view)
           content_view.check_ready_to_publish!
+
+          if options[:repos_units].present?
+            valid_labels_from_cv = content_view.repositories.map { |repo| repo.label }
```
Comment (ehelms) [inline app/lib/actions/katello/content_view/publish.rb]: Minor but you can shorten this to `content_view.repositories.map(&:label)`

**Comment by beav** [comment]
> Are we planning to handle errata_ids in a follow up PR? I'd appreciate an ACK from @jlsherrill on this given his expertise in the area.

Right now you auto-get the errata based on the RPMs selected. For example if you asked for "walrus-0.71" and there was an erratum with walrus-0.71 in it, you'd get that erratum.

We can pull in based on `errata_id` instead of `package_id` but I'd rather do it as a new PR.

**Comment by jlsherrill** [comment]
yeah, i'm okay with errata ids being a future feature

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 13]
```diff
@@ -81,11 +81,18 @@ def update
                                                            "in the content view version")
     param :major, :number, :desc => N_("Override the major version number"), :required => false
     param :minor, :number, :desc => N_("Override the minor version number"), :required => false
+
+    param :repos_units, Array, :desc => N_("Specify the list of units in each repo"), :required => false
```
Comment (jlsherrill) [inline app/controllers/katello/api/v2/content_views_controller.rb]: Can you better describe this param? 

I think you'd want something like:
```
param :repos_units, Array, :required => false do 
  param :rpm_filenames, Array, :required => false, :Desc => N_('some description')
end
```

**Comment by jlsherrill** [comment]
one comment, otherwise looks good to me

File: app/controllers/katello/api/v2/content_views_controller.rb (hunk 13 above)
Comment (beav) [reply app/controllers/katello/api/v2/content_views_controller.rb]: Ah I see, for some reason I thought you couldn't name inner params on arrays (which makes no sense).

Fixing now.

File: app/controllers/katello/api/v2/content_views_controller.rb [hunk 14]
```diff
@@ -81,11 +81,25 @@ def update
                                                            "in the content view version")
     param :major, :number, :desc => N_("Override the major version number"), :required => false
     param :minor, :number, :desc => N_("Override the minor version number"), :required => false
+
+    param :repos_units, Array, :desc => N_("Specify the list of units in each repo"), :required => false do
+      param :repo_units, Hash, :desc => N_("a hash containing a repo label and list of units"), :required => true do
+        param :label, String, :desc => N_("repo label"), :required => true
+        param :rpm_filenames, Array, :desc => N_("list of rpm filenames to include in published version"), :required => true do
+          param :rpm_filename, Array, :desc => N_("rpm filename"), :required => true
```
Comment (ehelms) [inline app/controllers/katello/api/v2/content_views_controller.rb]: I think this may confuse things since `rpm_filename` is not actually a param. And given `rpm_filenames` is an array of strings.

Comment (beav) [reply app/controllers/katello/api/v2/content_views_controller.rb]: I see, I can just say "list of rpm filename strings" for `:rpm_filenames` above and delete this line.

**Comment by beav** [comment]
```
Error:
Katello::DockerMetaTagTest#test_import_meta_tags:
NoMethodError: undefined method `schema2' for nil:NilClass
    /var/lib/workspace/workspace/katello-pr-test/test/models/docker_meta_tag_test.rb:58:in `test_import_meta_tags'
```

re-running since I made a small change anyway

**Comment by beav** [comment]
obtained ack from @zjhuntin as well, merging
